### ISSUE TITLE
False Positive with CS8604 and observables

### ISSUE THREAD

**Issue by Evangelink** [dotnet/roslyn#43498]
**Version Used**: 
- VS Version 16.5.4
- Lang version 8.0
- Net framework 4.8

**Steps to Reproduce**:

```csharp
IObserable<FileInfo?> myObs = GetSomething();
myObs.Where(x => x != null).Select(x => x.Name); // FP in x.Name
myObs.OfType<FileInfo>().Select(x => x.Name); // no FP in x.Name
```

**Expected Behavior**:
I would expect the code to have learnt from the `Where(x => x != null)`.

**Actual Behavior**:
`OfType` is ok but `Where(x => x != null)` is causing FP.

**Comment by CyrusNajmabadi** [comment]
This is not a Roslyn issue.  The compiler is warning as appropriate based on hte language rules so far.  If you'd like the language to change, could you port this over as a proposal to dotnet/csharplang?  Ideally it would be good if you could sketch out how the compiler would figure this out given that the signature for `.Where` is simply:   `public static IEnumerable<T> Where<T>(this IEnumerable<T>, Func<T, bool> predicate)`.  We'd need some way to say "if this predicate returns false on nulls that the result will contain no nulls".  Not really sure what's a good way to do that.

**Comment by Evangelink** [comment]
I was thinking about tweaking the dataflow analysis to be able to say whatever is used after this check has the not-null state.

**Comment by CyrusNajmabadi** [comment]
@Evangelink But how/why would dataflow analysis do this?  imagine i write a method called `DoesNotReturnValuesThatDoNotMatch`, and i call `.DoesNotReturnValuesThatDoNotMatch(x => x != null)`. how would it know that you're getting an `IEnumerable<FileInfo>` and not an `IEnumerable<FileInfo?>`

**Comment by CyrusNajmabadi** [comment]
Regardless, def open the csharplang issue.  We can brainstorm there.  for example, just saying "any lambda that does a null-check and returns that value will change the null state"
